dpc/src/virtual_machine/builder.rs

```diff
@@ -172,13 +170,14 @@ impl<N: Network> ResponseBuilder<N> {
         }
 
         // Ensure the value balance matches the fee from the request.
-        if value_balance != request.fee() {
-            return Err(anyhow!(
+        ensure!(
+            value_balance == request.fee(),
+            format!(
```

Comment:
I think you can avoid the `format!` and use instead:

```suggestion
        ensure!(
            value_balance == request.fee(),
            "Value balance does not match fee amount from request. Expected {} from request, found {} from response",
            request.fee(),
            value_balance
            );
```